### PR TITLE
Feature/more diagnostics bad blocks

### PR DESCRIPTION
**Pull request by LukaszRozmej**
## Changes

- Adds RLP dump option and adds it to default
- Adds extra data to log when block is invalid

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by MarekM25** [comment]
@LukaszRozmej I would consider adding RLP directly to the info log. It would improve the debugging

**Comment by LukaszRozmej** [comment]
> @LukaszRozmej I would consider adding RLP directly to the info log. It would improve the debugging

But then would it be good to be Info and default? Not sure if we want multiple rlps being logged on bad blocks. Could scare users.

**Comment by MarekM25** [comment]
> > @LukaszRozmej I would consider adding RLP directly to the info log. It would improve the debugging
> 
> But then would it be good to be Info and default? Not sure if we want multiple rlps being logged on bad blocks. Could scare users.

What about adding extra dump mode? It would be super useful now

File: src/Nethermind/Nethermind.Consensus/Validators/BlockValidator.cs [hunk 1]
```diff
@@ -129,53 +131,60 @@ public bool ValidateSuggestedBlock(Block block)
...
             for (int i = 0; i < processedBlock.Transactions.Length; i++)
             {
                 if (receipts[i].Error is not null && receipts[i].GasUsed == 0 && receipts[i].Error == "invalid")
                 {
-                    if (_logger.IsError) _logger.Error($"- invalid transaction {i}");
+                    _logger.Error($"- invalid transaction {i}");
                 }
             }
+
+            _logger.Error($"- parent beacon block root : expected {suggestedBlock.Header.ParentBeaconBlockRoot}, got {processedBlock.Header.ParentBeaconBlockRoot}");
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Consensus/Validators/BlockValidator.cs:182]: Screenshot from 2023-10-21 11-33-27
line to remove?

File: src/Nethermind/Nethermind.Blockchain/LogTraceDumper.cs [hunk 2]
```diff
@@ -11,25 +11,33 @@
...
+        Block block,
+        ILogger logger)
+    {
+        Keccak blockHash = block.Hash;
+        string fileName = $"block_{blockHash}.txt";
+        using FileStream diagnosticFile = GetFileStream(fileName);
+        Rlp rlp = new BlockDecoder().Encode(block, RlpBehaviors.AllowExtraBytes);
+        diagnosticFile.Write(rlp.Bytes);
+        if (logger.IsInfo)
+            logger.Info($"Created a Receipts trace of block {blockHash} in file {diagnosticFile.Name}");
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Blockchain/LogTraceDumper.cs:33]: "Created a Receipts trace of block" -> "Created a RLP dump of block"

File: src/Nethermind/Nethermind.Blockchain/LogTraceDumper.cs [hunk 3]
```diff
@@ -11,25 +11,33 @@
...
 {
-    public static List<JsonConverter> Converters { get; } = new List<JsonConverter>();
+    public static List<JsonConverter> Converters { get; } = new();
+
+    public static void LogDiagnosticRlp(
+        Block block,
+        ILogger logger)
+    {
+        Keccak blockHash = block.Hash;
+        string fileName = $"block_{blockHash}.txt";
```
Comment (rubo) [inline src/Nethermind/Nethermind.Blockchain/LogTraceDumper.cs:28]: The `.txt` extension is misleading as the file is binary. Let's either remove it or change to `.rlp`:

```suggestion
        string fileName = $"block_{blockHash}.rlp";
```

**Review by rubo — CHANGES_REQUESTED** [review]

**Review by rubo — APPROVED** [review]

**Pull request merged**